/**
 * Generated by @skyleague/therefore@v1.0.0-local
 * Do not manually touch this
 */
/* eslint-disable */

import type { DefinedError, ValidateFunction } from 'ajv'

import { validate as BulkProcessStatusValidator } from './schemas/bulk-process-status.schema.js'
import { validate as ErrorResponseValidator } from './schemas/error-response.schema.js'

/**
 * Holds information about the error encountered processing an item in bulk
 */
export interface BulkProcessingError {
    /**
     * Numeric value associated to the error reason
     */
    errorCode?: number | undefined
    /**
     * Error reason
     */
    errorReason?: string | undefined
    /**
     * Details about the error
     */
    errorSource?: string | undefined
    /**
     * Optional field populated only when request payload contains an externalId
     */
    externalId?: string | undefined
    /**
     * The index of the entity/item from bulk request that failed on processing
     */
    indexInRequest?: number | undefined
}

/**
 * Holds details about successful processed item
 */
export interface BulkProcessingSuccess {
    /**
     * Optional field populated only when request payload contains an externalId
     */
    externalId?: string | undefined
    /**
     * Unique identifier for the newly created resource
     */
    id?: string | undefined
    /**
     * The index of the entity/item from bulk request that failed on processing
     */
    indexInRequest?: number | undefined
}

/**
 * Holds information about the status of a bulk process
 */
export interface BulkProcessStatus {
    /**
     * List of errors
     */
    errors?: BulkProcessingError[] | undefined
    /**
     * List of successful processed items
     */
    processedItems?: BulkProcessingSuccess[] | undefined
    /**
     * Bulk process status
     */
    status?:
        | 'QUEUED'
        | 'IN_PROGRESS'
        | 'COMPLETE'
        | 'NOT_FOUND'
        | 'CANCEL'
        | 'TO_BE_CANCELED'
        | 'TIMED_OUT'
        | 'ERROR'
        | 'TRANSIENT_ERROR'
        | 'OVERRIDDEN'
        | 'RECOVERABLE_ERROR'
        | undefined
}

export const BulkProcessStatus = {
    validate: BulkProcessStatusValidator as ValidateFunction<BulkProcessStatus>,
    get schema() {
        return BulkProcessStatus.validate.schema
    },
    get errors() {
        return BulkProcessStatus.validate.errors ?? undefined
    },
    is: (o: unknown): o is BulkProcessStatus => BulkProcessStatus.validate(o) === true,
    parse: (o: unknown): { right: BulkProcessStatus } | { left: DefinedError[] } => {
        if (BulkProcessStatus.is(o)) {
            return { right: o }
        }
        return { left: (BulkProcessStatus.errors ?? []) as DefinedError[] }
    },
} as const

export interface ErrorResponse {
    errors?: RestError[] | undefined
}

export const ErrorResponse = {
    validate: ErrorResponseValidator as ValidateFunction<ErrorResponse>,
    get schema() {
        return ErrorResponse.validate.schema
    },
    get errors() {
        return ErrorResponse.validate.errors ?? undefined
    },
    is: (o: unknown): o is ErrorResponse => ErrorResponse.validate(o) === true,
    parse: (o: unknown): { right: ErrorResponse } | { left: DefinedError[] } => {
        if (ErrorResponse.is(o)) {
            return { right: o }
        }
        return { left: (ErrorResponse.errors ?? []) as DefinedError[] }
    },
} as const

export interface RestError {
    errorCode?: number | undefined
    errorReason?: string | undefined
    errorSource?: string | undefined
}
